File: codeforces/leq1000/733d.py
# D. Secret Santa
from collections import defaultdict
t = int(input())
def solve():
    ans = [0, []]
    n = int(input())
    wish = list(map(int, input().split()))
    sets = defaultdict(list)
    taken = set()
    free = set(range(1, n+1))
    for i, w in enumerate(wish, start=1):
        sets[w].append(i)
        if w in free:
            free.remove(w)
            ans[0] += 1
    fr2 = list(free)
    for w in wish:
        if w not in taken:
            ans[1].append(w)
            taken.add(w)
        else:
            ans[1].append(fr2.pop(-1))
    # print(ans[1])
    for i, w in enumerate(ans[1], start=1):
        if i == w:
            # print(sets[wish[i-1]])
            for x in sets[wish[i-1]]:
                if x != i:
                    # print(i-1, x-1)
                    ans[1][i-1], ans[1][x-1] = ans[1][x-1], ans[1][i-1]
                    break
    return ans

for _ in range(t):
    s = solve()
    print(f"{s[0]}\n{' '.join(map(str, s[1]))}")

File: codeforces/leq1000/733d.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> solve() {
    int n; cin >> n;
    vector<int> ans(n+1), wish(n);
    for (int& i : wish) cin >> i;
    unordered_map<int, vector<int>> sets;
    set<int> taken, free;

    for (int z = 1; z <= n; ++z)
        free.insert(z);
    for (int z = 1; z <= n; ++z) {
        int w = wish[z];
        sets[w].push_back(z);
        if (free.count(w)) {
            free.erase(w);
            ans[0] += 1;
        }
    }
    set<int, greater<int>> fr2;
    for (auto i : free) fr2.insert(i);
    for (auto w : wish) {
        if (!taken.count(w)) {
            ans.push_back(w);
            taken.insert(w);
        }
        else {
            auto x = fr2.begin();
            ans.push_back(*x);
            fr2.erase(x);
        }
    }
    for (int i = 1; i <= n; ++i) {
        int w = ans[i];
        if (i == w) {
            for (auto xx : sets[wish[i-1]]) {
                if (xx != i) {
                    swap(ans[i-1], ans[xx-1]);
                }
                break;
            }
        }

    }
    return ans;
}

int main() {
    int t; cin >> t;
    while (t--) {
        vector<int> v = solve();
        cout << v[0] << endl;
        for (int i = 1; i < v.size(); ++i) {
            cout << v[i] << " ";
        }
        cout << endl;
    }

}
